### ISSUE TITLE
Prismatik HDR bug - Light blinking when moving mouse 

### ISSUE THREAD

**Issue by zatagi** [psieg/Lightpack#255]
I have the issue here: https://www.youtube.com/watch?v=NgSb_JKcttU
FYI the controller is CH340 chinese one.

**Comment by MegaVat30** [comment]
Same here. I have used  ATmega328 CH340 Chinese replica and 2-3 days from installing everything was working normally, but suddenly the LEDs start flickering with different colors when I move the mouse. And it only happens when any window or app are running in fullscreen.

**Comment by psieg** [comment]
Is it related to HDR? Does it occur if HDR is disabled?

**Comment by zatagi** [comment]
@psieg Yes, it's HDR related since it's does not happened on SDR. My workaround is to have mouse trail enabled, I guess it's linked to refresh rate.

**Comment by sblantipodi** [comment]
I have the same problem here:
https://github.com/psieg/Lightpack/issues/355

**Comment by sblantipodi** [comment]
the problem happen if your grab interval is "too fast" for what your PC can capture.

If you can capture at 30FPS, set the grab interval under 33ms, 
if you can capture at 60FPS, set the grab interval under 16ms...

this is not a solution but a workaround, it should work.

**Comment by zomfg** [comment]
Try running [this build](https://github.com/zomfg/Lightpack/releases/download/5.11.2.18-grab-alpha/Prismatik_5.2.11.23_win_capture.zip) with flashes and without.
It should save frames in your `%tmp%`. Hopefully we'll be able to see what's going on.
Save frames in between sessions as it'll override the previous ones (don't run long sessions, depending on your disk space)

**Comment by sblantipodi** [comment]
I see nothing strange in the image produced by your build zomfg.

Imho the problem isn't in the capturing step but in the output of the software from the PC to the microcontroller.
Most probably it's not even a problem of the software that sends data to the microcontroller but simply in the transmission that isn't synced well.

I was able to reproduce the problem on a software written by my own. 

the problem is that some "dirty transmission" isn't blocked by the checksum check you use before displaying the colors on the led strip.
It's not a big problem thoug, it happen every three or four minutes and it last the time of a lightning.

**Comment by zomfg** [comment]
if by checksum you mean the 3 bytes after "Ada", that only applies to the led count, the color values aren't taken into account
if you set logs to 2, that'll log the color values the are sent to the device, if you can capture it while running something like [this](https://www.youtube.com/watch?v=xAwB9lQnxAY), maybe we'll see some pattern...

**Comment by jespertheend** [comment]
I'm pretty sure this is unrelated to sending data to the microcontroller. I am using Prismatik without any external ledstrip connected to my computer. Instead I use the socket api that comes with Prismatik to send the color values to my own application, which then sends http requests to my lights. I'm seeing the same issue without any microcontroller for a ledstrip.
In fact, if you look at the device tab, you can see the colors turning black sometimes.

Or at least, that's how it used to be for me anyway. I can't seem to replicate it anymore now, I recently switched from an amd card to one from nvidia, so perhaps that's related.

**Comment by sblantipodi** [comment]
Don't know what happened but I have the same problem of the opener. Prismatik is unusable when HDR is on

**Comment by MegaVat30** [comment]
In my case, the problem was solved by replacing all LED strips and wiring.
Apparently I had some flat LED strips.

**Comment by sblantipodi** [comment]
> In my case, the problem was solved by replacing all LED strips and wiring.
> Apparently I had some faulty LEDs.

I don't think that this can be the case since it works ok without HDR

**Comment by sblantipodi** [comment]
Ok I finally release my implementation of Screen Capture for Adalight devices and I don't have any problems with HDR.

Obviously I don't post the link to my implementation because it would not be correct to you guys, I love your community and your work here. If someone is interested in having a working HDR can try my implementation on my GitHub.

I still remain a big fan of prismatik and hope that this problem will be solved in the future.

PS: Zomfg, sorry I had no time to do that try, I will do it as soon as I can.

**Comment by Benik3** [comment]
Hello.
I'm also interested in HDR fix. I have now HDR 144Hz monitor. When switching to 10bit support or opening game with HDR I have this blinking.  
I tried to play with the source codes. I successfully updated dxgi to use FLIP SWAP CHAIN. Then I tried to change DXGI_FORMAT to R16G16B16A16_FLOAT without success. With R10G10B10A2_UNORM I have no flickering, but I got weird colors on the LEDs :/
DXGI and C++ is absolutely new to me, so it's hard to try something :D

**Comment by zomfg** [comment]
interesting...
could you push this to a branch on your fork?
this issue is very confusing since the documentation says you'll always get 8bit data, so I assumed it would shift rgb30 to rgb24 before passing data...
with HDR on, could you capture and save 1 raw frame with stock/8bit DDupl grabber, and 1 with your 10bit change
with something like this (change file path)
make half of your screen red/0xff0000 (fullscreen would be best), don't use a video source, make an image or html page
maybe do a couple of other colors, just note the hex values
thanks
```diff
diff --git a/Software/grab/GrabberBase.cpp b/Software/grab/GrabberBase.cpp
index 56a0b571..87278bf5 100644
--- a/Software/grab/GrabberBase.cpp
+++ b/Software/grab/GrabberBase.cpp
@@ -28,6 +28,7 @@
 #include "GrabberBase.hpp"
 #include "src/debug.h"
 #include <cmath>
+#include <QFile>
 
 namespace
 {
@@ -140,6 +141,7 @@ void GrabberBase::grab()
 		++grabScreensCount;
 		_context->grabResult->clear();
 
+		const GrabbedScreen *gs = NULL;
 		for (int i = 0; i < _context->grabWidgets->size(); ++i) {
 			if (!_context->grabWidgets->at(i)->isAreaEnabled()) {
 				_context->grabResult->append(qRgb(0,0,0));
@@ -149,6 +151,7 @@ void GrabberBase::grab()
 			getValidRect(widgetRect);
 
 			const GrabbedScreen *grabbedScreen = screenOfRect(widgetRect);
+			gs = grabbedScreen;
 			if (grabbedScreen == NULL) {
 				DEBUG_HIGH_LEVEL << Q_FUNC_INFO << " widget is out of screen " << Debug::toString(widgetRect);
 				_context->grabResult->append(0);
@@ -222,7 +225,9 @@ void GrabberBase::grab()
 				preparedRect);
 			_context->grabResult->append(avgColor);
 		}
-
+		QFile frameFile("/tmp/frame.dat");
+		frameFile.open(QIODevice::WriteOnly);
+		frameFile.write(QByteArray::fromRawData((const char*)gs->imgData, gs->imgDataSize));
 	}
 	emit frameGrabAttempted(_lastGrabResult);
 }

```

**Comment by Benik3** [comment]
I will try it.  
Here you can find my updated repository:  
https://github.com/Benik3/Lightpack/tree/HDR

**Comment by Benik3** [comment]
OK, I tried saving the frame as you mentioned.
First interesting thing, it doesn't happen with colors with full value (e.g. R255 G0 B0, but also R255 G255 B0 ...).  
With R10G10B10 on SDR the grab it see it like this:
When I try full screen "dirty" yellow (R 0xFF G0xF2 B0), when LEDs are really yellow there is 00F2FFFF, but when the LEDs changes blinks to red, there is FF2F0FC0. It looks like some problem between RGB and BGR.
When enabling 10bit in drivers, it see only FF2F0FC0.  
Is true that normally Prismatik use **B**8G8R8, but 10bit and 16bit format is only in order R10G10B10 (R16G16B16).

**Comment by Benik3** [comment]
I think that I found some trace.  
Always when the LEDs blinks, it's when AcquireNextFrame returns DXGI_ERROR_WAIT_TIMEOUT.
Otherwise I think that the 10bit DXGI_FORMAT is not the way, because when we use the 8bit, the dxgi should simply convert it and that's what we wont (LEDs are 8bit anyway)

**Comment by zomfg** [comment]
[re-reading remarks here](https://docs.microsoft.com/en-us/windows/win32/api/dxgi1_5/nf-dxgi1_5-idxgioutput5-duplicateoutput1) now makes more sense
so according to this the current grabber should produce 8888 colors no matter what, and upgrading the API only allows for bypassing the 8888 conversion from whatever
and yes, we don't want to convert 1010102 (or anything) ourselves because it's probably already very efficient as is
but someone reported weird colors with 10bit mode iirc, so just to be sure it would be nice to look at what's captured
you could alter my code snippet from above with something like this instead of `QByteArray` stuff
```cpp
QString path = QStandardPaths::writableLocation(QStandardPaths::TempLocation);
QImage imog(gs->imgData, gs->bytesPerRow / 4, gs->imgDataSize / gs->bytesPerRow, QImage::Format_RGB32);
imog.save(QString("%1/priz_%2_og.png").arg(path).arg(grabScreensCount), "PNG");
```
try this with `master` build (so old API), and the new API with everything 8888, and check if images look weird I guess (but look closely, it might be subtle)

now why would `AcquireNextFrame` timeout all of a sudden while moving mouse?... try looking into #373, maybe it's related

**Comment by Benik3** [comment]
True, after re-reading the remarks I understand it also more.  
I also investigated more the `AcquireNextFrame`. For better view I replaced the if statement with `returnBlackBuffer`. It looks like it's opposite. When it returns timeout, the picture is OK.  
What is weird, if I have window with the "yellow" pattern and I move mouse over it, it's grabbed like the "red", but when I catch the window and move with the window, it's right grabbed like yellow.  
Maybe it's some problem with Full screen grabbing and that the picture is in 8bit, but desktop maybe be in 10bit?  
Anyway what is second investigation, that with that `returnBlackBuffer` I got white flashes e.g. when I'm in the visual studio.

I will try tomorrow more :)

**Comment by Benik3** [comment]
Good vs bad captured image:  
![priz_26_og](https://user-images.githubusercontent.com/4532647/99395694-17fad800-28e1-11eb-8211-537f15fd4e3a.png)
![priz_27_og](https://user-images.githubusercontent.com/4532647/99395700-19c49b80-28e1-11eb-88eb-4d05684330d6.png)

**Comment by zomfg** [comment]
> Good vs bad captured image:

what was the test setup here? old api? new? 8bit?..

> And it looks like the `DXGI_ERROR_WAIT_TIMEOUT` is not really connected with it (black is when error is returned)

so it's both, crazy colors and timeouts (I don't have either in SDR)
can you check the captured `desc.Format` for each frame, in case crazy colors spit out something different
if they all are `DXGI_FORMAT_B8G8R8A8_UNORM` that's bad news (API bug?)

> BTW for some reason the saved pictures are wider then they should be (the black bar on the right). It happens even with `DownscaleMipLevel = 0` (picture has 3456 instead of 3440 px)

this is normal, it's just padding for memory alignment

**Comment by Benik3** [comment]
It was on the new API 8bit (in Prismatik) monitor is in 10bit.  
For sure I tested some DDupl sample app and didn't notice this.

**Comment by Benik3** [comment]
On old API the result looks same:
![priz_268_og](https://user-images.githubusercontent.com/4532647/99418022-db3bda80-28fa-11eb-94e9-4a4fd8800bcd.png)
![priz_269_og](https://user-images.githubusercontent.com/4532647/99418027-dc6d0780-28fa-11eb-9788-e52be6d8d694.png)

When I use new API and R10G10B10A2 DXGI format, it always looks this ugly, so it looks like it doesn't convert the image to 8bit in some cases?

**Comment by zomfg** [comment]
that's what I'd suspect yeah, and if `desc.Format` says 8888 we have no way of knowing that it's borked I guess

have you tried [this one](https://github.com/microsoftarchive/msdn-code-gallery-microsoft/tree/master/Official%20Windows%20Platform%20Sample/DXGI%20desktop%20duplication%20sample/%5BC%2B%2B%5D-DXGI%20desktop%20duplication%20sample/C%2B%2B)?
if that works consistently, try moving things in Prismatik til it works... or in the sample til it's broken in same way :D

**Comment by Benik3** [comment]
I just tried to write` desc.Format` and `texDesc.Format` to txt file and is true, that this value changes.  
Sometimes it returns 24 (10bit) and sometimes 87 (but isn't that DXGI_FORMAT_B5G5R5A1_UNORM?). It looks like that that's the problem. When it returns 24, the image is broken.

**Comment by zomfg** [comment]
what if you set `scaledTextureDesc.Format` to 8888?

**Comment by Benik3** [comment]
If I set `scaledTextureDesc.Format` it still flickers (but is interesting that saved png of `screen.imgData` is OK), but when I forced also `screen.imgFormat = mapDXGIFormatToBufferFormat(DXGI_FORMAT_B8G8R8A8_UNORM);` it looks working OK :)
Changing also `texture.Format` cause black image.

**Comment by zomfg** [comment]
so no issues now?..
you still get black frames?

**Comment by zomfg** [comment]
does this work?
```cpp
// ...
texDesc.Format = DXGI_FORMAT_B8G8R8A8_UNORM;
// ...
scaledTextureDesc.Format = texDesc.Format;
// ...
screen.imgFormat = mapDXGIFormatToBufferFormat(texDesc.Format);
// ...
```

**Comment by Benik3** [comment]
I just tried to play with it (no HDR, but 10bit ON) and it looks good :)
What works:
```
texDesc.Format = desc.Format;
scaledTextureDesc.Format = DXGI_FORMAT_B8G8R8A8_UNORM;
screen.imgFormat = mapDXGIFormatToBufferFormat(DXGI_FORMAT_B8G8R8A8_UNORM);
```
I will test more to see, if there are not any other issues.

**Comment by zomfg** [comment]
but does my snippet not work at all? I was hoping it would also cover the case without downscaling

black frames could be a combination of #373 and your grab interval (which is?) and refresh rate and/or maybe the limits of ddupl
try adjusting `ACQUIRE_TIMEOUT_INTERVAL`
and check if `returnBlackBuffer()` is called and from where

**Comment by Benik3** [comment]
Nope, it's doesn't work, resp. it shows black whenever there is 10bit image.  
After digging into it I found, that `CopyResource` and `CopySubresourceRegion` can be used only on same format type group. So both should be R8 or R10, R16... etc. That's probably why it doesn't work when it tries copy DDupl data to texture, but for some reason it works when it makes `CopySubresourceRegion` to `scaledTextureDesc`. Maybe because Both types (R8 and R10) are 32bit per pixel. See Remarks here: https://docs.microsoft.com/en-us/windows/win32/api/d3d11/nf-d3d11-id3d11devicecontext-copyresource
But it will not most likely works with R16G16B16A16 format (if someone will have it). So it's not 100% fix.

It looks like only proper fix is to convert the format (probably through pixel shader?), when source is not R8. But I think that's something I'm not able to program.

About grabbing I have now 10ms grab interval. I will try to test more.

**Comment by Benik3** [comment]
I tested more the flickering and I found, that the flickering is not black, but it's a previous frame...
![obrazek](https://user-images.githubusercontent.com/4532647/99569066-75218700-29d0-11eb-8ceb-5dc2336811eb.png)

**Comment by zomfg** [comment]
I tried these two
```cpp
texDesc.Format = desc.Format;
scaledTextureDesc.Format = DXGI_FORMAT_R8G8B8A8_UNORM;
screen.imgFormat = mapDXGIFormatToBufferFormat(DXGI_FORMAT_B8G8R8A8_UNORM);
```
```cpp
texDesc.Format = desc.Format;
scaledTextureDesc.Format = DXGI_FORMAT_R10G10B10A2_UNORM;
screen.imgFormat = mapDXGIFormatToBufferFormat(DXGI_FORMAT_B8G8R8A8_UNORM);
```
so with a working conversion I should get funky colors if I save those as `B8G8R8A8`, right?
but I get black frame every time

**Comment by Benik3** [comment]
hmm probably there is problem in copy from R8 to R10 texture? As I wrote is even weird that the copy texture copy from R10 to R8 works for me, because it shouldn't (from Microsoft description)...

**Comment by zomfg** [comment]
but even just BGR->RGB 8bit fails, so not even 8->10
while in your case both channel swap and depth change work apparently

**Comment by Benik3** [comment]
Ah, I see, weird

**Comment by Benik3** [comment]
Maybe this could help https://www.manongdao.com/q-1240529.html

**Comment by Benik3** [comment]
Maybe `ID3D11DeviceContext::OMSetRenderTargets method ` could be helpful, from remark: 

> Each render target can have a different data format. These render target formats are not required to have identical bit-per-element counts.

**Comment by zomfg** [comment]
probably
[ms ddupl sample uses it](https://github.com/microsoftarchive/msdn-code-gallery-microsoft/blob/master/Official%20Windows%20Platform%20Sample/DXGI%20desktop%20duplication%20sample/%5BC%2B%2B%5D-DXGI%20desktop%20duplication%20sample/C%2B%2B/OutputManager.cpp)

**Comment by Benik3** [comment]
Just for info, I tried to turn on windows HDR and grabbing doesn't work even after this "hack" :/  
I didn't investigate it more jet. But I was thinking, wouldn't be easier to user QImage:ConvertFormat on the final buffer? I will maybe try it.

**Comment by zomfg** [comment]
does not work as in you get all black? or random colors?
doing the conversion once the data is out of the texture would eat the cpu in a measurable way, especially when y'all get HDR 5k 240Hz and crank the grab interval 😄 so if it's avoidable...

**Comment by Benik3** [comment]
It's white, but I found that problem is in the pointer to img data (imog exist only in this function....).  
Well If I got it working, I can measure how much CPU it will eat. Also I'm converting the downscaled picture and grabbing over let say 90FPS I don't see as "must have". You will anyway start to hit possibilities of COM port (on wifi maybe not) and LED strips.
So better to have something which at least is able to grab correctly and slower then nothing :D

**Comment by Benik3** [comment]
Ok I got it working and with 12ms grabbing I have up to 15% CPU usage (i7-7600K 4,6GHz OC) :D

**Comment by Benik3** [comment]
OMG I thought that I have it working after all, but the QIamge doesn't have format for R16G16B16A16 (RGB64 doesn't fit) :(

**Comment by zomfg** [comment]
you mean `desc.Format` is R16... now?

**Comment by Benik3** [comment]
yep. That's when you switch windows to HDR (and maybe also in other HDR applications, but I didn't test it yet)

**Comment by Shaggy-One** [comment]
Confirming the "pointer trails" fix at least for desktop. Haven't had time to test in games yet but with Windows HDR and Win8DDupl in Prismatik enabled the flashing went away with "pointer trails" on at minimum setting in internet browsing.

Had the same issue as the title. All LEDs in the string flashing bright while mouse is moving with HDR and Win8 Desktop Duplication enabled. LEDs acting just fine in videos or games for the most part with the rare flash of "lightning".

Current version of Prismatik (PSIEG build 5.11.2.27), Arduino w/FASTLED (not sure what version of FASTLED as I lost source code years ago) and 138 leds using APA102 LEDs at 77hz capture rate. Windows HDR enabled running at 3440x1440 144hz, 10 bit YCbCr444 color set via NVIDIA control panel.

**Comment by nhac-lly** [comment]
Mouse trail is a good workaround because it forced the Grab Frequency to jump to maximum level at a quicker rate and not jumping up and down.
I'm thinking it will great if we can force it to a fixed number, it's another workaround but better than mouse trail I guess.

**Comment by EddoWagt** [comment]
For me the issue still occurs, even with HDR disabled. Pointer trails are really annoying to me... Ambibox 2.1.7 doesn't have this issue, but I have different issues with that right now

**Comment by goooroooX** [comment]
Is there any fix expected for this issue? 
"Win8 Desktop Duplication" mode still suffer from LEDs flickering when mouse is moving and HDR is enabled in Windows settings (Prismatik 5.11.2.27 + Adalight/Ambilight + Windows 10 build 19042.804).

**Comment by Benik3** [comment]
Nope, I tried some thing, but I didn't get it working on windows HDR (R16G16B16) with my skills.  
But this is opensource project, anyone can participate :)

**Comment by coltono** [comment]
Is this issue impossible to fix? 

This problem has to the most annoying part of using the program with peace & love.  Is there any way a non-developer like myself can contribute to help with this?

**Comment by psieg** [comment]
https://github.com/psieg/Lightpack/releases/tag/5.11.2.31 should fix this, thanks @romkal

**Comment by dormancygrace** [comment]
thanks! Works like a charm!

**Comment by Benik3** [comment]
The mouse is really fixed, thanks!  
Still there is a little problem with HDR, that the grabbed picture is over brightened when captured as SDR. But it is probably related with colour space  or something like this...  
![obrazek](https://user-images.githubusercontent.com/4532647/148663503-1f3172ed-261f-4d49-ac58-736866477889.png)

**Comment by jespertheend** [comment]
That would be #343.
Honestly I wouldn't be surprised if that turned out to be unfixable. Virtually any software that captures the screen is suffering from blown out colors. TeamViewer, Chrome Remote Desktop and SteamVR are all having the same issue when HDR is enabled.

**Comment by Benik3** [comment]
Yeah. I tried capture even with IDXGIOutput6 but that's not the problem :/  
This must be fixed by Microsoft probably.
